go/engine/saltpack_user_keyfinder.go

```diff
@@ -0,0 +1,270 @@
+		if err != nil {
+			return err
+		}
+		e.AddUserRecipient(m, &selfUpk.Current)
+	}
+
+	err = e.lookupRecipients(m)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// lookupRecipients adds the KID corresponding to each recipient to the recipientMap
+func (e *SaltpackUserKeyfinder) lookupRecipients(m libkb.MetaContext) error {
+	for _, u := range e.Arg.Recipients {
+		err := e.LookupUser(m, u) // For existing users
+		if err == nil {
+			continue
+		} else if _, isIdentifyFailedError := err.(libkb.IdentifyFailedError); !isIdentifyFailedError {
+			return err
+		}
+		return fmt.Errorf("Cannot find keys for %v: it is not an assertion for a registered user (err = %v)", u, err)
+	}
+	return nil
+}
+
+func (e *SaltpackUserKeyfinder) LookupUser(m libkb.MetaContext, user string) error {
+
+	Arg := keybase1.Identify2Arg{
+		UserAssertion: user,
+		Reason: keybase1.IdentifyReason{
+			Type: keybase1.IdentifyReasonType_ENCRYPT,
+		},
+		AlwaysBlock:      true,
+		IdentifyBehavior: keybase1.TLFIdentifyBehavior_CLI,
+	}
+	eng := NewResolveThenIdentify2(m.G(), &Arg)
+	if err := RunEngine2(m, eng); err != nil {
+		return libkb.IdentifyFailedError{Assertion: user, Reason: err.Error()}
+	}
+
+	engRes := eng.Result()
+	if engRes == nil {
+		return fmt.Errorf("Null result from Identify2")
+	}
+	arg := libkb.NewLoadUserByUIDArg(m.Ctx(), e.G(), engRes.Upk.GetUID())
+	upak, _, err := e.G().GetUPAKLoader().LoadV2(arg)
+	if err != nil {
+		return err
+	}
+
+	return e.AddUserRecipient(m, &upak.Current)
+}
+
+func (e *SaltpackUserKeyfinder) hasRecipientDeviceOrPaperKeys(id keybase1.UID) bool {
+	_, ok := e.RecipientDeviceAndPaperKeyMap[id]
+	return ok
+}
+
+func (e *SaltpackUserKeyfinder) hasRecipientEntityKeys(id keybase1.UserOrTeamID) bool {
+	_, ok := e.RecipientEntityKeyMap[id]
+	return ok
+}
+
+func (e *SaltpackUserKeyfinder) AddUserRecipient(m libkb.MetaContext, upk *keybase1.UserPlusKeysV2) error {
+	if err := e.AddDeviceAndPaperKeys(m, upk); err != nil {
+		return err
+	}
+	return e.AddPUK(m, upk)
+}
+
+func (e *SaltpackUserKeyfinder) AddDeviceAndPaperKeys(m libkb.MetaContext, upk *keybase1.UserPlusKeysV2) error {
+	if !e.Arg.UsePaperKeys && !e.Arg.UseDeviceKeys {
+		// No need to add anything
+		return nil
+	}
+
+	if e.hasRecipientDeviceOrPaperKeys(upk.Uid) {
+		// This user's keys were already added
+		return nil
+	}
+
+	var keys []keybase1.KID
+	hasPaperKey := false
+	hasDeviceKey := false
+
+	hasPUK := len(upk.PerUserKeys) > 0
+
+	for KID, key := range upk.DeviceKeys {
+		// Note: for Nacl encryption keys, the DeviceType field is not set, so we need to look at the "parent" signing key
+		if libkb.KIDIsDeviceEncrypt(KID) && key.Parent != nil && libkb.IsPaperKey(upk.DeviceKeys[*key.Parent]) {
+			hasPaperKey = true
+			if e.Arg.UsePaperKeys {
+				keys = append(keys, KID)
+				// TODO Is it ok to put recipient names in the logs?
+				e.G().Log.CDebugf(m.Ctx(), "adding user %v's paper key", upk.Username)
```

Comment:
`m.CDebugf`

Reply:
yes, it's inferred from everything around there...  the logs are private unless the user `keybase log send`s